selfdrive/ui/qt/onroad.cc

```diff
@@ -174,43 +174,41 @@ void OnroadAlerts::paintEvent(QPaintEvent *event) {
   }
 }
 
-
-ExperimentalButton::ExperimentalButton(QWidget *parent) : QPushButton(parent) {
+// ExperimentalButton
+ExperimentalButton::ExperimentalButton(QWidget *parent) : experimental_mode(false), QPushButton(parent) {
   setVisible(false);
   setFixedSize(btn_size, btn_size);
-  setCheckable(true);
 
-  params = Params();
   engage_img = loadPixmap("../assets/img_chffr_wheel.png", {img_size, img_size});
   experimental_img = loadPixmap("../assets/img_experimental.svg", {img_size, img_size});
+  QObject::connect(this, &QPushButton::clicked, this, &ExperimentalButton::changeMode);
+}
 
-  QObject::connect(this, &QPushButton::toggled, [=](bool checked) {
-    params.putBool("ExperimentalMode", checked);
-  });
+void ExperimentalButton::changeMode() {
+  auto params = Params();
+  const auto cp = (*uiState()->sm)["carParams"].getCarParams();
+  const bool experimental_mode_available = cp.getExperimentalLongitudinalAvailable() ? params.getBool("ExperimentalLongitudinalEnabled") : cp.getOpenpilotLongitudinalControl();
+  bool can_change = experimental_mode_available && params.getBool("ExperimentalModeConfirmed");
+  if (can_change) {
+    params.putBool("ExperimentalMode", !experimental_mode);
+  }
```

Comment:
Can we make a global variable or helper function to consolidate all of this experimental/alpha longitudinal logic? I think we have very similar logic in settings.cc as well as here which is pretty confusing to keep track of

Reply:
How about implementing it in another PR? 
Think it would be better to make a more comprehensive implementation instead of just targeting this specific functionality.For example, adding a member variable "struct settings" to `uistate` to track all changes in settings. This way, most calls to the `param.read` in the UI can be eliminated. It also allows for real-time switching of states, such as between miles and kilometers.